### PR TITLE
Disable multiple kernel instantiations when using HIP

### PR DESCRIPTION
**Pull request by Rombur**
Related to #4546

### CODE REVIEW

File: cmake/kokkos_enable_options.cmake [hunk 1]
```diff
@@ -57,6 +57,7 @@ KOKKOS_ENABLE_OPTION(TUNING               OFF "Whether to create bindings for tu
 KOKKOS_ENABLE_OPTION(AGGRESSIVE_VECTORIZATION OFF "Whether to aggressively vectorize loops")
 KOKKOS_ENABLE_OPTION(LAUNCH_COMPILER      ON  "Whether to potentially use the launch compiler")
 KOKKOS_ENABLE_OPTION(COMPILE_AS_CMAKE_LANGUAGE OFF "Whether to use native cmake language support")
+KOKKOS_ENABLE_OPTION(HIP_MULTIPLE_INST    OFF "Wether multiple kernels are instantiated at compile time - improve performance but increase compile time")
```
Comment (dalg24) [inline cmake/kokkos_enable_options.cmake:60]:
```suggestion
KOKKOS_ENABLE_OPTION(HIP_MULTIPLE_INST    OFF "Whether multiple kernels are instantiated at compile time - improve performance but increase compile time")
```

**Comment by dalg24** [comment]
OpenMPTarget failure is clearly unrelated
```
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestUniqueToken.hpp:151: Failure
4: Expected equality of these values:
4:   sum
4:     Which is: 9999990
4:   int64_t(N) * R
4:     Which is: 10000000
4: [  FAILED  ] openmptarget.unique_token_instance (261 ms)
```
@crtrott @rgayatri23 is that something worth investigating?

**Comment by rgayatri23** [comment]
> OpenMPTarget failure is clearly unrelated
> 
> ```
> 4: /var/jenkins/workspace/Kokkos/core/unit_test/TestUniqueToken.hpp:151: Failure
> 4: Expected equality of these values:
> 4:   sum
> 4:     Which is: 9999990
> 4:   int64_t(N) * R
> 4:     Which is: 10000000
> 4: [  FAILED  ] openmptarget.unique_token_instance (261 ms)
> ```
> 
> @crtrott @rgayatri23 is that something worth investigating?

There is no implementation for the instance scope of UniqueToken yet. Only the global scope is implemented. 
I thought that this test case was blocked for OpenMPTarget backend.

**Review by dalg24 — APPROVED** [review]

File: cmake/kokkos_enable_options.cmake [hunk 2]
```diff
@@ -57,6 +57,7 @@ KOKKOS_ENABLE_OPTION(TUNING               OFF "Whether to create bindings for tu
 KOKKOS_ENABLE_OPTION(AGGRESSIVE_VECTORIZATION OFF "Whether to aggressively vectorize loops")
 KOKKOS_ENABLE_OPTION(LAUNCH_COMPILER      ON  "Whether to potentially use the launch compiler")
 KOKKOS_ENABLE_OPTION(COMPILE_AS_CMAKE_LANGUAGE OFF "Whether to use native cmake language support")
+KOKKOS_ENABLE_OPTION(HIP_MULTIPLE_KERNEL_INSTANTIATIONS OFF "Whether multiple kernels are instantiated at compile time - improve performance but increase compile time")
```
Comment (masterleinad) [inline cmake/kokkos_enable_options.cmake:60]:
```suggestion
KOKKOS_ENABLE_OPTION(HIP_MULTIPLE_KERNEL_INSTANTIATIONS OFF "Whether multiple versions for each kernel are instantiated at compile time - improve performance but increase compile time")
```

**Comment by dalg24** [comment]
Both HIP builds are passing.  Failures are unrelated (issues with one of the testing machine)

**Pull request merged**